common/client/node_lifecycle.go

```diff
@@ -226,47 +211,151 @@ func (n *node[CHAIN_ID, HEAD, RPC]) aliveLoop() {
+			receivedNewHead := n.onNewFinalizedHead(lggr, &localHighestChainInfo, latestFinalized)
+			if receivedNewHead && noNewFinalizedBlocksTimeoutThreshold > 0 {
+				finalizedHeadsSub.ResetTimer(noNewFinalizedBlocksTimeoutThreshold)
 			}
-
-			latestFinalizedBN := latestFinalized.BlockNumber()
-			if latestFinalizedBN > localHighestChainInfo.FinalizedBlockNumber {
-				promPoolRPCNodeHighestFinalizedBlock.WithLabelValues(n.chainID.String(), n.name).Set(float64(latestFinalizedBN))
-				localHighestChainInfo.FinalizedBlockNumber = latestFinalizedBN
+		case <-finalizedHeadsSub.NoNewHeads:
+			// We haven't received a finalized head on the channel for at least the
+			// threshold amount of time, mark it broken
+			lggr.Errorw(fmt.Sprintf("RPC's finalized state is out of sync; no new finalized heads received for %s (last finalized head received was %v)", noNewFinalizedBlocksTimeoutThreshold, localHighestChainInfo.FinalizedBlockNumber), "latestReceivedBlockNumber", localHighestChainInfo.BlockNumber)
+			if n.poolInfoProvider != nil {
+				if l, _ := n.poolInfoProvider.LatestChainInfo(); l < 2 {
+					lggr.Criticalf("RPC's finalized state is out of sync; %s %s", msgCannotDisable, msgDegradedState)
+					// We don't necessarily want to wait the full timeout to check again, we should
+					// check regularly and log noisily in this state
+					finalizedHeadsSub.ResetTimer(zombieNodeCheckInterval(noNewFinalizedBlocksTimeoutThreshold))
+					continue
+				}
 			}
+			n.declareOutOfSync(syncIssueNoNewFinalizedHead)
+			return
+		case <-finalizedHeadsSub.Errors:
+			lggr.Errorw("Finalized heads subscription was terminated", "err", err)
+			n.declareUnreachable()
+			return
 		}
 	}
 }
 
-func (n *node[CHAIN_ID, HEAD, RPC]) isOutOfSync(num int64, td *big.Int) (outOfSync bool) {
-	outOfSync, _ = n.syncStatus(num, td)
-	return
+type headSubscription[HEAD any] struct {
+	Heads      <-chan HEAD
+	Errors     <-chan error
+	NoNewHeads <-chan time.Time
+
+	noNewHeadsTicker *time.Ticker
+	sub              types.Subscription
+	cleanUpTasks     []func()
+}
+
+func (sub *headSubscription[HEAD]) ResetTimer(duration time.Duration) {
+	sub.noNewHeadsTicker.Reset(duration)
+}
+
+func (sub *headSubscription[HEAD]) Unsubscribe() {
+	for _, doCleanUp := range sub.cleanUpTasks {
+		doCleanUp()
+	}
+}
+
+func (n *node[CHAIN_ID, HEAD, PRC]) registerNewSubscription(ctx context.Context, lggr logger.SugaredLogger,
+	noNewDataThreshold time.Duration, newSub func(ctx context.Context) (<-chan HEAD, types.Subscription, error)) (headSubscription[HEAD], error) {
+	result := headSubscription[HEAD]{}
+	var err error
+	var sub types.Subscription
+	result.Heads, sub, err = newSub(ctx)
+	if err != nil {
+		return result, err
+	}
+
+	result.Errors = sub.Err()
+	lggr.Debug("Successfully subscribed")
+
+	// TODO: will be removed as part of merging effort with BCI-2875
+	result.sub = sub
+	//n.stateMu.Lock()
+	//n.healthCheckSubs = append(n.healthCheckSubs, sub)
+	//n.stateMu.Unlock()
```

Comment:
Are these a placeholder?

Reply:
Yes, after merging the BCI-2875 `result.sub = sub` will be replaced with 
```
n.stateMu.Lock()
n.healthCheckSubs = append(n.healthCheckSubs, sub)
n.stateMu.Unlock()
```